### ISSUE TITLE
DeepSpeedEngineWrapper.backward() does a bit too much

### ISSUE THREAD

**Issue by zhc7** [huggingface/accelerate#2951]
The source code of DeepSpeedEngineWrapper:
```
class DeepSpeedEngineWrapper:
    """
    Internal wrapper for deepspeed.runtime.engine.DeepSpeedEngine. This is used to follow conventional training loop.

    Args:
        engine (deepspeed.runtime.engine.DeepSpeedEngine): deepspeed engine to wrap
    """

    def __init__(self, engine):
        self.engine = engine

    def backward(self, loss, **kwargs):
        # runs backpropagation and handles mixed precision
        self.engine.backward(loss, **kwargs)

        # Deepspeed's `engine.step` performs the following operations:
        # - gradient accumulation check
        # - gradient clipping
        # - optimizer step
        # - zero grad
        # - checking overflow
        # - lr_scheduler step (only if engine.lr_scheduler is not None)
        self.engine.step()
        # and this plugin overrides the above calls with no-ops when Accelerate runs under
        # Deepspeed, but allows normal functionality for non-Deepspeed cases thus enabling a simple
        # training loop that works transparently under many training regimes.
```

My question is: Why do we need to do `self.engine.step()` here immediately? This behavior zeros grad and change the parameter without noticing the user. It might be out of expectation. Since backward step is internally binded with zeroing grad and changing parameter, this blocks users from checking the gradient or parameter manually before stepping.

I know deepspeed-wrapped models can't be seen as normal models, but this behavior still elimiates a lot of flexibility.

**Comment by muellerzr** [comment]
Hitting this snag right now actually, will see what we decide to do

**Comment by muellerzr** [comment]
We're working with the DS team to try and remove the `engine` entirely, however as a user you can always call `model.engine.backward()` etc manually without harm in accelerate

**Comment by nom** [comment]
Somehow 
```
loss = loss / accelerator.gradient_accumulation_steps
accelerator.deepspeed_engine_wrapped.engine.backward(loss)
```

is not equivalent to
accelerator.backward(loss)                    

with deepspeed. What gives?

**Comment by LinB203** [comment]
same question here.

**Comment by GreenWindow1997** [comment]
same question here. It's indeed very hard to understand why it was designed this way. I believe accelerator.backward(loss) should only perform the backward operation, and other steps should be written outside this function in a more standard and understandable manner.

**Comment by LinB203** [comment]
Not stale.

**Comment by Nirmal-Adhikari-hub** [comment]
`self.engine.module.named_parameters()` 
might have the gradients you are looking for. I am not sure about this as I am not the person who knows enough to answer these esoteric questions but I also went through a similar problem of wanting to observe the gradients of the model after the backward pass and found out this solution which worked for me.

I hope it helps and if it did glad to be of any help to you.

Happy Coding!

**Comment by muellerzr** [comment]
So for a bit more context, we've been waiting on *this* PR to happen, so hopefully we can give more flexibility soon: https://github.com/deepspeedai/DeepSpeed/pull/7018

**Comment by MaxHeuillet** [comment]
Hello, any update regarding this issue?

**Comment by hzphzp** [comment]
any update?

**Comment by hzphzp** [comment]
Any update?

**Comment by Boltzmachine** [comment]
Does that mean if I use deepspeed integration with huggingface's `Trainer`, the parameters will be updated twice in each step?

**Comment by naomili0924** [comment]
> Does that mean if I use deepspeed integration with huggingface's `Trainer`, the parameters will be updated twice in each step?

no, it won't.

**Comment by naomili0924** [comment]
https://github.com/huggingface/accelerate/pull/3819  I created this PR to debug parameters for a short time, feel free to use it. But even with it still heavily depends on the deepspeed engine.

**Comment by longspringcity** [comment]
same question here.
